### ISSUE TITLE
Ackerman Perturbation behavior

### ISSUE THREAD

**Issue by FrancescoMandru** [lyft/l5kit#299]
I'm looking at the Ackerman Perturbation and I'm visualizing some samples.  I don't know if it's desired behavior but sometimes the images are rotated, while this doesn't happen without the Ackerman perturbation. This results in trajectories not aligned with the road.  Which trajectories are modified? Past, Future or both?

![download (1)](https://user-images.githubusercontent.com/43784964/121714029-c369b500-cadd-11eb-91f6-9eb77b7dfaba.png)

Moreover I wonder if it is a normal behavior that when I add the Ackerman Perturbation the training of my network it becomes very very slow. I tought it was applied to the dataset before the training . Thank you!

**Comment by lucabergamini** [comment]
> Moreover I wonder if it is a normal behavior that when I add the Ackerman Perturbation the training of my network it becomes very very slow. I tought it was applied to the dataset before the training . Thank you!

Yes this is expected (you should also see a higher loss) because perturbations are online data augmentation, so fitting your training data will take longer.



> Which trajectories are modified? Past, Future or both?

100% sure about future, I'll check for past



> I'm looking at the Ackerman Perturbation and I'm visualizing some samples. I don't know if it's desired behavior but sometimes the images are rotated

Do you have the code to reproduce this? Please ensure you fix the random seed so we can reproduce it. Also, try to use as less custom code as possible so that we can pin the issue down to this specific module

**Comment by FrancescoMandru** [comment]
This is the dataset generation function:

```python
def datasetGeneration(dataset_path, 
                      cfg):
    ## BASE ROOT OF DATASET USED FOR LOCAL DATA MANAGER
    os.environ["L5KIT_DATA_FOLDER"] = dataset_path

    ## INITIALLIZE THE CLASS OF THE PATHS
    dm = LocalDataManager()

    # BUILD THE RASTERIZER
    rasterizer = build_rasterizer(cfg, dm)

    ## RETUNR THE PATH REQUIRED
    train_path = dm.require('scenes/train.zarr')

    ## OPEN THE ZARR FILE
    train_zarr = ChunkedDataset(train_path).open()

    ## DEFINE THE PERTURBATION
    perturb_prob = cfg["train_data_loader"]["perturb_probability"]
    mean = np.array([0.0, 0.0, 0.0])  # lateral, longitudinal and angular
    std = np.array([0.5, 1.5, np.pi / 6])
    perturbation = AckermanPerturbation(
            random_offset_generator=GaussianRandomGenerator(mean=mean, std=std), perturb_prob=perturb_prob)

    ## BUILD TRAIN DATALOADER PERTURBED
    train_dataset_p = EgoDataset(cfg, train_zarr, rasterizer, perturbation)
    train_dataset_p = TransformDataset(train_dataset_p, cfg)


    ## BUILD TRAIN DATALOADER
    train_dataset = EgoDataset(cfg, train_zarr, rasterizer)
    train_dataset = TransformDataset(train_dataset, cfg)
    
    return train_dataset_p, train_dataset
```

The Transform Dataset function is the same used also in the previous issue of the Trajectory rotation. Then I plot my results also like in the previous issue, which work perfectly when the trajectory is not perturbed

```python
image = train_dataset_p[index][0]
image = image.transpose(1, 2, 0)
image = np.sum(image,axis=2)

target_train_ack = train_dataset_p[index][1] #index is a single sample of my dataset, 1 are the target positions
history_train_ack = train_dataset_p[index][3] #index is a single sample of my dataset, 3 are the past positions

W = cfg['raster_params']['raster_size'][0]
# agent's position (in pixels)
agent_position_image = np.array([cfg['raster_params']['ego_center'][0] * W,
                                 cfg['raster_params']['ego_center'][1] * W])

# transform to pixels by division on r and then translate
r = cfg['raster_params']['pixel_size'][0]

positions_pixels_history_ack = agent_position_image + history_train_ack/r
positions_pixels_future_ack = agent_position_image + target_train_ack/r
positions_pixels_history_ack = np.clip(positions_pixels_history_ack, a_min=(0,0), a_max=(W, W))
positions_pixels_future_ack = np.clip(positions_pixels_future_ack, a_min=(0,0), a_max=(W, W))


#############
#  Plotting
#############

plt.figure(figsize=(5,5))

plt.imshow(image)

plt.scatter(positions_pixels_future_ack[:, 0],
             positions_pixels_future_ack[:, 1],
             s=4,
             marker='s',
             c='#33CC00')

plt.scatter(positions_pixels_history_ack[:, 0],
            positions_pixels_history_ack[:, 1],
            s=3,
            c='#33CC00')

plt.tight_layout()
plt.axis('off')

plt.show()
```

This code produce the first image when the Ackerman perturbation is not applied (with probability 0.5) and the second/third when it is applied:
![download](https://user-images.githubusercontent.com/43784964/122090766-6dac4a00-ce08-11eb-8971-cec9c1c60a60.png)
![download (1)](https://user-images.githubusercontent.com/43784964/122090786-71d86780-ce08-11eb-8567-e23b671ce396.png)
![download (2)](https://user-images.githubusercontent.com/43784964/122090789-71d86780-ce08-11eb-989a-5f1fdc58dd55.png)

From these images I can agree with you that only future positions are modified. But for example if the Ackerman perturbation change the groun truth trajectory to be outside of the road, should it be marked as a bad behavior?

**Comment by lucabergamini** [comment]
Sorry but this code is not really reproducible on my end.

If this is an issue with the Ackerman perturbation you should be able to see it just by:
- loading the data;
- fixing the seed;
- creating an EgoDataset with perturbation enabled;
- visualise some data with the default visualizer (`to_rgb` method)

Otherwise we can't pin it down to the Ackerman module (it may be an issue in `TransformDataset`, in your visualisation, etc..)

**Comment by FrancescoMandru** [comment]
Hi @lucabergamini . Seems it was a problem on my side with visualization. I think that it is fixed now. Anyway how can i save in png the resulting image after applying the `draw_trajectory` method? It seems that standard saving methods doesn't save the image correctly.
